rust/passwords/src/cbc.rs

```diff
@@ -0,0 +1,162 @@
+    0x40, 0x40, 0x43, 0x43, 0x45, 0x45, 0x46, 0x46, 0x49, 0x49, 0x4A, 0x4A, 0x4C, 0x4C, 0x4F, 0x4F,
+    0x51, 0x51, 0x52, 0x52, 0x54, 0x54, 0x57, 0x57, 0x58, 0x58, 0x5B, 0x5B, 0x5D, 0x5D, 0x5E, 0x5E,
+    0x61, 0x61, 0x62, 0x62, 0x64, 0x64, 0x67, 0x67, 0x68, 0x68, 0x6B, 0x6B, 0x6D, 0x6D, 0x6E, 0x6E,
+    0x70, 0x70, 0x73, 0x73, 0x75, 0x75, 0x76, 0x76, 0x79, 0x79, 0x7A, 0x7A, 0x7C, 0x7C, 0x7F, 0x7F,
+];
+
+/// Convert the key to the proper format to give to CBC Encryptor and Decryptor.
+fn derive_key(pw: &[u8]) -> [u8; 24] {
+    let mut result = SEED;
+    for (idx, &b) in pw.iter().enumerate() {
+        result[idx & 7] ^= b;
+    }
+
+    let mut k8 = [0u8; 8];
+    for i in 0..8 {
+        k8[i] = PARITY_BITS[(result[i] & 0x7F) as usize];
+    }
+
+    let mut key_24 = [0u8; 24];
+    key_24[0..8].copy_from_slice(&k8);
+    key_24[8..16].copy_from_slice(&k8);
+    key_24[16..24].copy_from_slice(&k8);
+    key_24
+}
+
+pub fn cbc_encrypt(key: &[u8], data: &[u8]) -> Result<Vec<u8>, CbcError> {
+    let hashed_key = derive_key(key);
+    let iv = [0u8; 8];
+    let padding_len = (8 - ((data.len() + 4) % 8)) % 8;
+
+    // ciphertext = ENC_SIG + bytes([padding * 16 + 0xE]) + data + bytes(padding)
+    let buf_len = ENC_SIG.len() + 1 + data.len() + padding_len;
+    let mut buf = Vec::with_capacity(buf_len);
+    buf.extend_from_slice(ENC_SIG);
+    buf.push((padding_len * 16 + 0xE) as u8);
+    buf.extend_from_slice(data);
+    buf.extend(std::iter::repeat_n(0, padding_len));
+
+    let cipher = cbc::Encryptor::<TdesEde3>::new(&hashed_key.into(), &iv.into());
+
+    let ct = cipher
+        .encrypt_padded_mut::<NoPadding>(&mut buf, buf_len)
+        .map_err(|_| CbcError::EncryptionFailed)?;
+
+    Ok(B64.encode(ct).into_bytes())
+}
+
+pub fn cbc_decrypt(key: &[u8], b64_encrypted_data: &[u8]) -> Result<Vec<u8>, CbcError> {
+    let hashed_key = derive_key(key);
+    let iv = [0u8; 8];
+    let mut ciphertext = B64
+        .decode(b64_encrypted_data)
+        .map_err(|_| CbcError::InvalidBase64)?;
+
+    let cipher = cbc::Decryptor::<TdesEde3>::new(&hashed_key.into(), &iv.into());
+
+    let pt = cipher
+        .decrypt_padded_mut::<NoPadding>(&mut ciphertext)
+        .map_err(|_| CbcError::DecryptionFailed)?;
+
+    // Validate ENC SIGN
+    if pt.len() < 4 || &pt[0..3] != ENC_SIG {
+        return Err(CbcError::InvalidSignature);
+    }
+
+    // Parse the Metadata byte (4th byte)
+    let meta_byte = pt[3];
+    if meta_byte < 0xE {
+        return Err(CbcError::InvalidSignature);
+    }
+    let padding_len = ((meta_byte as usize) - 0xE) / 16;
+
+    // Layout: [SIG (3)] [META (1)] [DATA (len - 4 - padding)] [NULLS (padding)]
+    let end_idx = pt.len() - padding_len;
+    if end_idx < 4 {
+        return Err(CbcError::InvalidSignature);
+    }
+
+    Ok(pt[4..end_idx].to_vec())
+}
+
+pub fn cbc_check_password(key: &[u8], ciphertext: &[u8]) -> bool {
+    cbc_decrypt(key, ciphertext).is_ok()
+}
+
+#[cfg(feature = "cbc")]
```

Comment:
I don't think you need this here, since the whole cbc module is declared under a feature.